pkg/kv/kvserver/rafttrace/rafttrace.go

```diff
@@ -0,0 +1,170 @@
+// Copyright 2024 The Cockroach Authors.
+//
+// Use of this software is governed by the CockroachDB Software License
+// included in the /LICENSE file.
+
+package rafttrace
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/cockroachdb/cockroach/pkg/raft"
+	"github.com/cockroachdb/cockroach/pkg/raft/raftpb"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+	"github.com/cockroachdb/cockroach/pkg/util/tracing"
+	"github.com/cockroachdb/cockroach/pkg/util/tracing/tracingpb"
+)
+
+type traceValue struct {
+	// ctx is a trace specific context used to log events on this trace.
+	ctx context.Context
+	// baseCtx is the underlying proposal buffer which we additionally log to.
+	baseCtx withContext
+	// cleanup will retrieve all events that were logged in this trace.
+	cleanup func() tracingpb.Recording
+}
+
+// RaftTracer is a utility to trace raft messages. Note that it may erroneously
+// include incorrect messages since it does not consider the term, however it is
+// easier to filter out extra messages that deal with missing messages.
+type RaftTracer struct {
+	// If this is too slow consider a bloom filter to see if this event is
+	// possibly traced. This map should normally be very small.
+	m syncutil.Map[uint64, traceValue]
```

Comment:
This `Map` approach does not take advantage of the fact that entries are ordered by `LogMark` in most places. We also can't just do point deletions from it: the `MsgStorageAppendResp` event is not scoped to a single entry, it rather says that all entries up to a `LogMark` are synced.

Reply:
The assumption is that the number of entries will typically be small enough that we don't need to keep them sorted to be efficient. Instead the code just iterates over all of them.